### PR TITLE
[docs] Initial Refine documentation

### PR DESCRIPTION
**Pull request by bsouthga**
Adds api documentation for refine under `/docs/api-reference/refine`. 

Pages added:

- `/docs/api-reference/refine/Refine`: Index / landing page with example + overviews
- `/docs/api-reference/refine/Checkers`: Overview of refine checkers
  - `/docs/api-reference/refine/Primitive_Checkers`
  - `/docs/api-reference/refine/Collection_Checkers`
  - `/docs/api-reference/refine/Advanced_Checkers`
-  `/docs/api-reference/refine/Utilities`: JSON parsing + assertion functions

### CODE REVIEW

**Comment by bsouthga** [comment]
@drarmstr -- implemented feedback, feel free to merge (or let me know when you'd like to do it)

File: docs/docs/api-reference/refine/Primitive_Checkers.md [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+const value: Date = result.value;
+
+// test an invalid value
+const failedResult = check(1);
+assert(failedResult.type === 'failure');
+```
+
+## `jsonDate()`
+
+Similar to date, though also will implicitly coerce ISO date strings to Date objects.
```
Comment (drarmstr) [inline docs/docs/api-reference/refine/Primitive_Checkers.md:152]: Probably useful to mention that this is particularly useful for synchronizing with JSON serialization of Date objects

File: docs/docs/api-reference/refine/Collection_Checkers.md [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+// test an invalid value
+assert(check('test').type === 'failure');
+assert(check({a: 'test', b: 1}).type === 'failure');
+assert(check({a: 1, c: 'test'}).type === 'failure');
+```
+
+## `set()`
+
+Checker to assert if a mixed value is a Set type with valid values.
```
Comment (drarmstr) [inline docs/docs/api-reference/refine/Collection_Checkers.md:96]: Checker to assert if a mixed value is a `Set` type with valid values.

File: docs/docs/api-reference/refine/Collection_Checkers.md [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+const value: $ReadOnlySet<number> = result.value;
+
+// test an invalid value
+assert(check('test').type === 'failure');
+assert(check({a: 'test', b: 'other', c: 3}).type === 'failure');
+```
+
+## `map()`
+
+Checker to assert if a mixed value is a Map type with valid keys and values.
```
Comment (drarmstr) [inline docs/docs/api-reference/refine/Collection_Checkers.md:116]: Checker to assert if a mixed value is a `Map` type with valid keys and values.

**Comment by bsouthga** [comment]
pushed additional nit fixes -- also did a sanity test of building docasaurus locally and everything looks fine.

**Pull request merged**